src/PowerShellEditorServices/Language/RunspaceSynchronizer.cs

```diff
@@ -0,0 +1,272 @@
+            targetEngineIntrinsics = ReflectionUtils.GetEngineIntrinsics(runspaceTarget);
+            IsReadyForEvents = true;
+
+            sourceEngineIntrinsics.Events.SubscribeEvent(
+                source: null,
+                eventName: null,
+                sourceIdentifier: PSEngineEvent.OnIdle.ToString(),
+                data: null,
+                handlerDelegate: HandleRunspaceStateChange,
+                supportEvent: true,
+                forwardEvent: false);
+
+            Activate();
+            // Trigger events
+            HandleRunspaceStateChange(sender: null, args: null);
+        }
+
+        /// <summary>
+        /// Does a thing
+        /// </summary>
+        public static void Activate()
+        {
+            SourceActionEnabled = true;
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static void HandleRunspaceStateChange(object sender, PSEventArgs args)
+        {
+            if (!SourceActionEnabled)
+            {
+                return;
+            }
+
+            SourceActionEnabled = false;
+
+            var newOrChangedModules = new List<PSModuleInfo>();
+            List<PSModuleInfo> modules = ReflectionUtils.GetModules(sourceRunspace);
+            foreach (PSModuleInfo module in modules)
+            {
+                if (moduleCache.Add(module))
+                {
+                    newOrChangedModules.Add(module);
+                }
+            }
+
+
+            var newOrChangedVars = new List<PSVariable>();
+
+            var variables = sourceEngineIntrinsics.GetVariables();
+            foreach (var variable in variables)
+            {
+                // If the variable is a magic variable or it's type has not changed, then skip it.
+                if(POWERSHELL_MAGIC_VARIABLES.Contains(variable.Name) ||
+                    (variableCache.TryGetValue(variable.Name, out Type value) && value == variable.Value?.GetType()))
+                {
+                    continue;
+                }
+
+                // Add the variable to the cache and mark it as a newOrChanged variable.
+                variableCache[variable.Name] = variable.Value?.GetType();
+                newOrChangedVars.Add(variable);
+            }
+
+            if (newOrChangedModules.Count > 0)
+            {
+                // Import the modules in the targetRunspace with -Force
+                using (PowerShell pwsh = PowerShell.Create())
+                {
+                    pwsh.Runspace = targetRunspace;
+
+                    foreach (PSModuleInfo moduleInfo in newOrChangedModules)
+                    {
+                        if(moduleInfo.Path != null)
+                        {
+                            pwsh.AddCommand("Import-Module")
+                                .AddParameter("Name", moduleInfo.Path)
+                                .AddParameter("Force")
+                                .AddStatement();
```

Comment:
This will only import the `psm1`, so it will miss nested/required modules, required assemblies, and any command filtering.  We also can't actually determine if the module info object was imported via the `psm1` directly or via the `psd1`, which could yield significantly different results.

Reply:
>This will only import the `psm1`

Can you elaborate on this? When a module is imported through a manifest, the `ModuleInfo.Path` points to the `psd1` file. For example, 
```
PS:34> $s = gmo Microsoft.PowerShell.Management
PS:35> $s.Path
F:\pscore70\Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1
```

Reply:
@daxian-dbw for some reason... for script modules, the path points to the `psm1` :( Patrick is right.

Reply:
Ah, I see. The `Microsoft.PowerShell.Management` is a bad example. The assembly is listed as a `NestedModule`, so that's literally a manifest module :) If the `RootModule` or `ModuleToProcess` is specified, then the path would point to the specified module file I guess.

A bit more logic is needed here to find out if there is a module manifest file.

Reply:
We can do something similar to:

```
Join-Path (Get-Module PowerShellGet).ModuleBase "$((Get-Module PowerShellGet).Name).psd1"
```

Reply:
Problem is you can't tell whether they've imported the `psm1` or the `psd1`.

One could, for instance import a `psm1` directly to get intellisense on functions that are typically filtered by the `psd1`.  Conversely if you always assume `psm1`, you miss out on nested/required modules/assemblies and type/format files.

Reply:
For the `psd1` vs. `psm1` case, we can tell which one was used in loading by inspecting the version. If you are importing a `psm1` file, the version would be `0.0`. For a `psd1`, it's very unlikely `0.0`.

Reply:
Alternatively, we could override Import-Module in order to get the exact parameters used for `Import-Module` and then run the real import-module in each runspace.

Reply:
I'm now importing the module base if the version is greater than `0.0`. overriding `Import-Module` was more work than it was worth.